### DISCUSSION TITLE
Is there a way to wrap theorems and the linke in tcolorboxes when the output is PDF?

### DISCUSSION THREAD

**Discussion by darthlite** [quarto-dev/quarto-cli#7313] [Q&A]
### Description

I have written a Quarto document with some theorems, proofs, examples, etc. When I compile to PDF, these naturally show up as formatted text. Is it possible (without changing the qmd file) to have the theorems and examples be outputted in tcolorboxes similar to the following picture:
![jrrdF](https://github.com/quarto-dev/quarto-cli/assets/116678368/34ea9702-9543-40af-81ce-324a0b0f0e75)

**Comment by mcanouil** [comment]
You would need to write a LUA filter to handle this.
See <https://quarto.org/docs/extensions/filters.html>.

**Reply by cscheid** [reply]
Like @mcanouil said, you'll have to write a Lua filter. But let me add a little more context.

In version 1.3, this isn't really possible because theorems are crossreferenceable, and that system wasn't extensible.

However, it is possible in 1.4. What you'll be doing is effectively creating a new rendering method for Theorem types (which are a new type of node in 1.4).

Let's start with a .qmd file:

````
---
title: Hello
format: pdf
filters:
  - theorem.lua
---

::: {#thm-line}

## Line

The equation of any straight line, called a linear equation, can be written as:

$$
y = mx + b
$$
:::

See @thm-line.
````

Now, let's write a sequence of progressively more complicated Theorem renderers. (More on further replies below)

**Reply by cscheid** [reply]
Here's the very simplest way to render a Theorem, which does effectively nothing except testing that we're actually changing the output:

<table>
<tr><td>

````lua
-- theorem.lua
quarto._quarto.ast.add_renderer("Theorem", function()
  return quarto._quarto.format.isLatexOutput()
end, function(thm)
  return pandoc.RawBlock("latex", "\\textbf{A THEOREM}")
end)

return {}
````
</td><td>
<img width="726" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/d55e5154-504c-43ff-ba02-9290160d8176">
</td></tr>
</table>

**Reply by cscheid** [reply]
Now let's write a Theorem almost like quarto does. (The actual quarto rendered for "Theorem" handles theorems, lemmas, corollaries, propositions, conjecture, definitions, examples, and exercises. We'll keep it simple here for now).

<table>
<tr><td>

````lua
-- theorem.lua
quarto._quarto.ast.add_renderer("Theorem", function()
  return quarto._quarto.format.isLatexOutput()
end, function(thm)
  -- these are the fields 
  -- this is a pandoc.Div containing the contents of the thm
  local el = thm.div
  -- this is a pandoc.Inlines({}) containing the name of the thm
  local name = quarto.utils.as_inlines(thm.name)
  local label = thm.identifier -- this is the label of the thm

  local preamble = pandoc.Para(
    pandoc.RawInline("latex", "\\begin{theorem}"))
  preamble.content:insert(pandoc.RawInline("latex", "["))
  if name then
    preamble.content:extend(name)
  end
  preamble.content:insert(pandoc.RawInline("latex", "]"))
  preamble.content:insert(pandoc.RawInline("latex",
    "\\protect\\hypertarget{" .. label .. 
    "}{}\\label{" .. label .. "}")
  )
  el.content:insert(1, preamble)
  el.content:insert(pandoc.Para(
    pandoc.RawInline("latex", "\\end{theorem}")))
  -- See #3776
  el.attr.identifier = ""
  return el
end)
return {}
````
</td><td width="50%">
<img width="722" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/b0c4d61a-cd56-41ea-af0c-5932e1b46cf4">
</td></tr>
</table>

**Reply by cscheid** [reply]
I've now done the easy part. From here, you'll need to learn how to make tcolorboxes that have latex's theorem crossreferencing information (which isn't at all trivial, and is probably easiest to learn by writing plain .tex files first). Then, you can use the template above to emit custom theorems however you see fit.

**Reply by darthlite** [reply]
Many thanks! I'll see if I can figure it out from here.
